src/cargo/util/lints.rs

```diff
@@ -665,13 +665,20 @@ pub fn unexpected_target_cfgs(
         return Ok(());
     };
 
-    if !global_check_cfg.exhaustive {
+    // If we have extra `--check-cfg` args comming from the lints config, we need to
+    // refetch the `--print=check-cfg` with those extra args.
+    let lint_rustflags = pkg.manifest().lint_rustflags();
+    let check_cfg = if lint_rustflags.iter().any(|a| a == "--check-cfg") {
```

Comment:
I don't feel too comfortable having this stringly-typed API from our `Cargo.toml` parser to here

Reply:
Me either. Looking at the flags like this is ugly, I would love to not have to do this. A better way would probably be to pass always pass `--print=check-cfg` (with a file path) in the check/build invocation of the crate, that way we only have one compilation unit. It would be a much larger change though.